PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,449 @@
+        module, as the values of this dictionary are lists. Hooks for the same
+        module will be run in the order in which they are cached. Previously
+        cached hooks are always preserved rather than overidden.
+
+        By default, official hooks are cached _before_ user-defined hooks. For
+        modules with both official and user-defined hooks, this implies that the
+        former take priority over and hence will be loaded _before_ the latter.
+
+        Parameters
+        ----------
+        hook_dir : str
+            Absolute or relative path of the directory containing **hook
+            scripts** (i.e., Python scripts with filenames matching
+            `hook-{module_name}.py`, where `{module_name}` is the module hooked
+            by that script) to be cached. For convenience, this method applies
+            tilde and variable expansion to this path (e.g., a leading `~` will
+            be replaced by the absolute path of this user's home directory).
+        """
+
+        # Canonicalize this directory's path and validate its existence.
+        hook_dir = os.path.abspath(expand_path(hook_dir))
+        if not os.path.isdir(hook_dir):
+            raise FileNotFoundError(
+                'Hook directory "{}" not found.'.format(hook_dir))
+
+        # For each hook script in this directory...
+        hook_filenames = glob.glob(os.path.join(hook_dir, 'hook-*.py'))
+        for hook_filename in hook_filenames:
+            # Fully-qualified name of this hook's corresponding module,
+            # constructed by removing the "hook-" prefix and ".py" suffix.
+            module_name = os.path.basename(hook_filename)[5:-3]
+
+            # Lazily loadable hook object.
+            module_hook = ModuleHook(
+                module_graph=self.module_graph,
+                module_name=module_name,
+                hook_filename=hook_filename,
+                hook_module_name_prefix=self._hook_module_name_prefix,
+            )
+
+            # If at least one hook object has been cached for this module,
+            # append this hook object to the existing list of such objects.
+            if module_name in self:
+                self[module_name].append(module_hook)
+            # Else, create a new list containing this hook object.
+            else:
+                self[module_name] = [module_hook]
+
+
+    def uncache_modules(self, module_names):
+        """
+        Remove the passed modules and hence all hook scripts cached for these
+        modules from this cache.
+
+        Parameters
+        ----------
+        module_names : list
+            List of all fully-qualified module names to be removed.
+        """
+
+        for module_name in set(module_names):  # Eliminate duplicate entries.
+            if module_name in self:
+                del self[module_name]
+
+
+#FIXME: For space efficiency, convert to use slots rather than "__dict__".
```

Comment:
Probably not worth the effort unless the hook files number in the thousands or more.